docs/apis/subsystems/roles.md

```diff
@@ -0,0 +1,141 @@
+
+- A **permission** is the level in which a capability is applied to a role. Example values include:
+  - Allow
+  - Prevent
+  - Prohibit
+  - Inherit
+- A **context** is an area in Moodle. There are several context types for:
+  - the whole site
+  - a user
+  - a course category
+  - a course
+  - an activity
+  - a block
+
+### Roles
+
+Role defines list of permissions - role definition is global for all contexts, but can be changed by local context overrides.
+
+Permissions control possible user actions within Moodle (eg delete discussions, add activities etc)
+
+Roles can be applied to users in a **context** (eg assign Fred as a teacher in a particular course)
+
+### Context
+
+Here are the possible contexts, listed from the most general to the most specific.
+
+| CONTEXT NAME        | CONTEXT AREA       | CONTEXT LEVEL |
+|---------------------|--------------------|---------------|
+| `CONTEXT_SYSTEM`    | the whole site     | 10            |
+| `CONTEXT_USER`      | another user       | 30            |
+| `CONTEXT_COURSECAT` | a course category  | 40            |
+| `CONTEXT_COURSE`    | a course           | 50            |
+| `CONTEXT_MODULE`    | an activity module | 70            |
+| `CONTEXT_BLOCK`     | a block            | 80            |
+
+An authorized user will be able to assign an arbitrary number of roles to each user in any context.
+
+See [Roles and modules#Context](https://docs.moodle.org/dev/Roles_and_modules#Context) for more information.
+
+### Capabilities
+
+Capabilities can have the following permissions:
+
+1. `CAP_INHERIT`
+2. `CAP_ALLOW`
+3. `CAP_PREVENT`
+4. `CAP_PROHIBIT`
+
+If no permission is defined, then the capability permission is inherited from a context that is more general than the current context. If we define different permission values for the same capability in different contexts, we say that we are overriding the capability in the more specific context.
+
+### Capability conflicts
+
+Since the capabilities in each role could be different, there could be conflict in capabilities.
+
+If we set a `PROHIBIT` on a capability, it means that the capability cannot be overridden and will ALWAYS  have a permission of prevent (deny). Prohibit always wins.   For example, Jeff has a naughty student role that prohibits him from postings in any forums (for the whole site), but he's also assigned a facilitator role in "Science forum" in the course Science and Math 101. Since prohibit always wins, Jeff is unable to post in "Science forum".
+
+## Programming Interface
+
+Although the Roles system may look complicated at first glance, implementing it in Moodle code is fairly simple.
+
+- Examples of functions handling the **Roles** can be found at [`accesslib.php`](https://github.com/moodle/moodle/blob/master/lib/accesslib.php)
+
+- Moodle comes with a list of predefined roles for example Student, Teacher, Course Manager etc. A new role can be created based on
+    those predefined roles using **role archetype:**
```

Comment:
```suggestion
- Moodle comes with a list of predefined roles, including a Student, Teacher, Non-editing teacher, and Course Manager role. Each of these roles are based on a **role archetype**, which acts as a template for roles. Any custom role created by the site administrator can also _choose_ to follow one of these role archetypes. When a plugin defines a new capability, it may specify how it would expect to be applied within these role archetypes, and these are applied to any role which follows this archetype. For example, if you create a new activity module with named _mod_example_, with a capability 'mod/example:view', you may specify that the `teacher`, and `editingteacher` archetypes are granted the capability with the allow permission. Any role which is based on these archetypes will be granted this capability with the 'allow' permission.
```